handler/projection/compactor_test.go

```diff
@@ -0,0 +1,120 @@
+package projection_test
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/dogmatiq/dodeca/logging"
+	"github.com/dogmatiq/dogma"
+	. "github.com/dogmatiq/dogma/fixtures"
+	. "github.com/dogmatiq/verity/handler/projection"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"golang.org/x/sync/semaphore"
+)
+
+var _ = Describe("type Compactor", func() {
+	var (
+		logger    *logging.BufferedLogger
+		handler   *ProjectionMessageHandler
+		compactor *Compactor
+	)
+
+	BeforeEach(func() {
+		logger = &logging.BufferedLogger{}
+		handler = &ProjectionMessageHandler{}
+		compactor = &Compactor{
+			Handler:   handler,
+			Interval:  1 * time.Millisecond,
+			Timeout:   1 * time.Millisecond,
+			Semaphore: semaphore.NewWeighted(1),
+			Logger:    logger,
+		}
+	})
+
+	Describe("func Run()", func() {
+		It("compacts immediately", func() {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			handler.CompactFunc = func(
+				context.Context,
+				dogma.ProjectionCompactScope,
+			) error {
+				cancel()
+				return nil
+			}
+
+			err := compactor.Run(ctx)
+			Expect(err).To(Equal(context.Canceled))
+		})
+
+		It("compacts repeatedly", func() {
+			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+			defer cancel()
+
+			called := true
+			handler.CompactFunc = func(
+				context.Context,
+				dogma.ProjectionCompactScope,
+			) error {
+				if called {
+					cancel()
+				}
+
+				called = true
+				return nil
+			}
+
+			err := compactor.Run(ctx)
+			Expect(err).To(Equal(context.Canceled))
+		})
+
+		It("returns if the context is canceled while waiting for the sempahore", func() {
```

Comment:
```suggestion
		It("returns if the context is canceled while waiting for the semaphore", func() {
```

Also, does this return an error or just returns?

Reply:
It returns the `context.Canceled` error.